File: aoc2022/kotlin/src/main/kotlin/Day02.kt
import util.AdventOfCode

object Day02 : AdventOfCode() {
    sealed class RpsMove(val points: Int) {
        fun willBeat(mine: RpsMove) = this.beats == mine
        fun willLoseTo(mine: RpsMove) = this.losesTo == mine

        val beats: RpsMove
            get() = when (this) {
                Rock -> Scissors
                Paper -> Rock
                Scissors -> Paper
            }

        val losesTo: RpsMove
            get() = when (this) {
                Rock -> Paper
                Paper -> Scissors
                Scissors -> Rock
            }

        object Rock : RpsMove(1)
        object Paper : RpsMove(2)
        object Scissors : RpsMove(3)
    }

    private fun String.toRpsMove() = when (this) {
        "A", "X" -> RpsMove.Rock
        "B", "Y" -> RpsMove.Paper
        "C", "Z" -> RpsMove.Scissors
        else -> throw IllegalArgumentException("`$this` is not a valid move.")
    }

    private fun score(theirs: RpsMove, mine: RpsMove) =
        mine.points + when {
            theirs.willBeat(mine) -> 0
            theirs.willLoseTo(mine) -> 6
            else -> 3
        }

    private fun scoreByResult(theirs: RpsMove, result: String) =
        when (result) {
            "X" -> theirs.beats.points
            "Y" -> 3 + theirs.points
            "Z" -> 6 + theirs.losesTo.points
            else -> throw IllegalArgumentException()
        }


    fun part1(input: List<String>) =
        input.map { line -> line.split(" ").map { it.toRpsMove() } }
            .sumOf { (theirs, mine) -> score(theirs, mine) }

    fun part2(input: List<String>) =
        input.map { it.split(" ") }
            .sumOf { (theirs, outcome) -> scoreByResult(theirs.toRpsMove(), outcome) }


    @JvmStatic
    fun main(args: Array<String>) {
        println("Day 2")
        println("\tPart 1: ${part1(inputFileLines)}")
        println("\tPart 2: ${part2(inputFileLines)}")
    }
}

File: aoc2022/fsharp/Days/Day02.fs
module Day02

open FSharpx

type Outcome =
    | Lose
    | Draw
    | Win

    member this.Points =
        match this with
        | Lose -> 0
        | Draw -> 3
        | Win -> 6

type Move =
    | Rock
    | Paper
    | Scissors

    member this.Points: int =
        match this with
        | Rock -> 1
        | Paper -> 2
        | Scissors -> 3

    member this.Beats: Move =
        match this with
        | Rock -> Scissors
        | Paper -> Rock
        | Scissors -> Paper

    member this.LosesTo: Move =
        match this with
        | Rock -> Paper
        | Paper -> Scissors
        | Scissors -> Rock

    member this.WillBeat(m: Move) : bool = this.Beats = m
    member this.WillLoseTo(m: Move) : bool = this.LosesTo = m

let toMove (m: string) =
    match m with
    | "A"
    | "X" -> Rock
    | "B"
    | "Y" -> Paper
    | "C"
    | "Z" -> Scissors
    | _ -> failwith $"`{m}` is not a valid move."

let toOutcome (m: string) =
    match m with
    | "X" -> Lose
    | "Y" -> Draw
    | "Z" -> Win
    | _ -> failwith $"`{m}` is not a valid outcome."

let computeOutcome (theirs: Move) (mine: Move) =
    match theirs, mine with
    | t, m when (t = m) -> Draw
    | t, m when t.WillBeat m -> Lose
    | t, m when t.WillLoseTo m -> Win
    | t, m -> failwith $"I don't know what to do with {t} and {m}"


let score (theirs: Move) (mine: Move) : int =
    (computeOutcome theirs mine).Points + mine.Points

let scoreByResult (theirs: Move) (result: Outcome) : int =
    let mine =
        match result with
        | Lose -> theirs.Beats
        | Draw -> theirs
        | Win -> theirs.LosesTo

    mine.Points + result.Points

let sumFn (txA: string -> 'A) (txB: string -> 'B) (transform: 'A -> 'B -> int) (line: string list) : int =
    match line with
    | a :: b :: _ -> transform (txA a) (txB b)
    | _ -> failwith $"Could not process line {line}"

let split (line: string) =
    line |> String.splitChar [| ' ' |] |> Array.toList

let part1 (input: string[]) =
    input |> Seq.map split |> Seq.sumBy (sumFn toMove toMove score)

let part2 (input: string[]) =
    input |> Seq.map split |> Seq.sumBy (sumFn toMove toOutcome scoreByResult)
